### PR TITLE
Add settings and favorites for wear so users can quickly execute actions they do regularly faster

### PR DESCRIPTION
**Pull request by dshokouhi**
## Summary

- Adds a new settings section so users can set and clear favorites (clear is important in case the entity list is old like the user deleted an entity)
- When favorites are set they load as a chip on the home screen, then as soon as entities are loaded they convert to toggles if necessary. This allows them to be executed before we finish loading

Fixes: #1881 
## Screenshots

image

image

image

image

image

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#610

## Any other notes

Theres a lot going on in this PR in a future PR lets do the phone side for grabbing and setting favorites :)

### CODE REVIEW

File: wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt [hunk 1]
```diff
@@ -103,10 +119,11 @@ class HomeActivity : ComponentActivity(), HomeView {
         finish()
     }
 
+    @ExperimentalWearMaterialApi
     @Composable
-    private fun LoadHomePage(entities: Array<Entity<Any>>) {
+    private fun LoadHomePage(entities: Array<Entity<Any>>, favorites: MutableSet<String>) {
 
-        if (entities.isNullOrEmpty()) {
+        if (entities.isNullOrEmpty() && favorites.isNullOrEmpty()) {
```
Comment (leroyboerefijn) [inline wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt:126]: Is there any possibility where entities is null or empty, but favorites is not?

File: wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt [hunk 2]
```diff
@@ -317,11 +384,190 @@ class HomeActivity : ComponentActivity(), HomeView {
...
+                    .padding(top = 10.dp),
+                icon = {
+                    Image(asset = CommunityMaterial.Icon.cmd_cog)
+                },
+                label = {
+                    Text(
+                        text = stringResource(id = R.string.settings)
+                    )
+                },
+                onClick = { swipeDismissableNavController.navigate("Settings") },
```
Comment (leroyboerefijn) [inline wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt:403]: Shouldn't you use SCREEN_SETTINGS here?

**Review by leroyboerefijn — COMMENTED** [review]
Awesome work once again! I'm impressed with your speed 😋

Haven't tested it out yet, but looks good!

File: wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt (hunk 1 above)
Comment (dshokouhi) [reply wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt:126]: Yes there is, so down below in the `else` we do individual checks for favorites and entities again. So basically if the user has not had entities loaded yet and hasnt set favorites they see old loading. If they dont have entities yet but do have favorites they will see favorites and the loading chip, then loading screen/chip will disappear once we get the list of entities.

File: wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt (hunk 2 above)
Comment (dshokouhi) [reply wear/src/main/java/io/homeassistant/companion/android/home/HomeActivity.kt]: lol yup totally missed it haha, fixed now

**Comment by leroyboerefijn** [comment]
I'm not sure what the best practices are for navigation in compose. But I was wondering if the HomeActivity is not becoming too large when it will contain all of the settings pages and subpages as well?
Would it make sense to have all settings related stuff in a separate Activity?

**Comment by dshokouhi** [comment]
Well we could move the composable to a separate file but navigating to composables instead of activities seems to be what the docs say.

https://developer.android.com/jetpack/compose/navigation

Also using the current method gives us the cool animation and handles the backstack etc..

**Comment by leroyboerefijn** [comment]
> Well we could move the composable to a separate file but navigating to composables instead of activities seems to be what the docs say.
> 
> https://developer.android.com/jetpack/compose/navigation
> 
> Also using the current method gives us the cool animation and handles the backstack etc..

Okay good then! I'll try to test it soon, am curious about the animations and functionality!

**Comment by dshokouhi** [comment]
ok did some refactoring so HomeActivity has a bit less going on, hopefully that helps make things easier to follow

File: wear/src/main/java/io/homeassistant/companion/android/settings/SettingsComposable.kt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                },
+                enabled = entityViewModel.favoriteEntities.isNotEmpty()
+            )
+        }
+    }
+
+    @Composable
+    fun ScreenSetFavorites(
+        validEntities: List<Entity<Any>>,
+        scalingLazyListState: ScalingLazyListState,
```
Comment (leroyboerefijn) [inline wear/src/main/java/io/homeassistant/companion/android/settings/SettingsComposable.kt:100]: I believe this should be a separate object. Now the state of the list is shared between lists. This also causes the set favorites list to not start at the top for me.

Comment (dshokouhi) [reply wear/src/main/java/io/homeassistant/companion/android/settings/SettingsComposable.kt:100]: hmm I think that its probably best to always start at the top here, maybe later on we can sort based on the checked status?

File: wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import com.mikepenz.iconics.typeface.library.community.material.CommunityMaterial
+import io.homeassistant.companion.android.R
+import io.homeassistant.companion.android.home.HomePresenter
+import io.homeassistant.companion.android.viewModels.EntityViewModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+
+class CommonFunctions {
```
Comment (JBassett) [inline wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt:27]: I don't think this really needs to be a class with a bunch of functions, rather just a bunch of top level functions.

File: wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        )
+    }
+
+    fun updateFavorites(entityViewModel: EntityViewModel, presenter: HomePresenter) {
+        mainScope.launch { entityViewModel.favoriteEntities = presenter.getWearHomeFavorites().toMutableSet() }
+    }
+
+    fun saveFavorites(favorites: Set<String>, presenter: HomePresenter) {
+        mainScope.launch { presenter.setWearHomeFavorites(favorites.toSet()) }
+    }
```
Comment (JBassett) [inline wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt:57]: Pass in the scope so you can make these top level.

File: wear/src/main/java/io/homeassistant/companion/android/settings/SettingsComposable.kt [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+import com.mikepenz.iconics.compose.Image
+import com.mikepenz.iconics.typeface.library.community.material.CommunityMaterial
+import io.homeassistant.companion.android.R
+import io.homeassistant.companion.android.common.data.integration.Entity
+import io.homeassistant.companion.android.home.HomeActivity
+import io.homeassistant.companion.android.home.HomePresenter
+import io.homeassistant.companion.android.util.CommonFunctions
+import io.homeassistant.companion.android.viewModels.EntityViewModel
+
+class SettingsComposable {
```
Comment (JBassett) [inline wear/src/main/java/io/homeassistant/companion/android/settings/SettingsComposable.kt:39]: Again if these are passing everything via function parameters should this be a class?

**Review by JBassett — CHANGES_REQUESTED** [review]

File: wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt (hunk 4 above)
Comment (dshokouhi) [reply wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt]: yea good point :)

File: wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt (hunk 5 above)
Comment (dshokouhi) [reply wear/src/main/java/io/homeassistant/companion/android/util/CommonFunctions.kt]: sure, thank you :)

File: wear/src/main/java/io/homeassistant/companion/android/settings/SettingsComposable.kt (hunk 6 above)
Comment (dshokouhi) [reply wear/src/main/java/io/homeassistant/companion/android/settings/SettingsComposable.kt]: almost missed this one 😛  thank you

**Review by JBassett — APPROVED** [review]

**Pull request merged**